### PR TITLE
Export Constructors

### PR DESCRIPTION
**Pull request by koehlma**
I am using this library for a [Python interpreter written in JavaScript](https://github.com/koehlma/jaspy). To make JS objects work together with the Python environment I have to patch the prototypes. Therefore I need access to the constructors.

### CODE REVIEW

**Comment by peterolson** [comment]
I don't have any big objections to this patch, but is there any reason why you don't just grab the constructors without modifying the library? For example, you could do this:

    var SmallInteger = bigInt().constructor,
        BigInteger = bigInt(1e16).constructor;

**Comment by koehlma** [comment]
I am doing something like this right now. But it looks more like a hack than a clean solution. Also in terms of API stability, the above behavior not part of the official API. Adding this to the API ensures that no one would accidentally break things because I am relying on implicitly given features.

**Comment by peterolson** [comment]
Doesn't exposing them to the public API also seem like a hack? The `BigInteger` and `SmallInteger` constructors are implementation details that I don't really intend end-users to deal with. Let's suppose in the future I want to get rid of these constructors because I found an alternative approach to do things behind the scenes that has better performance.

If I make these part of the public API, then I have to make a choice between breaking backward compatibility but improving performance, or keeping sub-optimal performance but maintaining backwards compatibility.

I don't have any current plans to replace the `BigInteger` and `SmallInteger` constructors, but I'm not going to make any guarantees that they're here to stay, either. Even if I made them part of the public API, I'm not sure I'm willing to promise that I'll keep them around for the sake of backwards compatibility. In principle, it seems like a bad idea to make hidden implementation details public, especially if I don't have any intention of maintaining their stability.

I see where you're coming from, but I don't think that exposing these constructors will make your code more future proof. How about this: if I have any plans to change the behavior of these constructors, I'll try to notify you in advance.

**Comment by Yaffle** [comment]
@peterolson 

```
    function BigInteger(value, sign) {
        this.value = value; // array or number
        this.sign = sign;
    }

    BigInteger.prototype.multiply = function (v) {
      var n = parseValue(v);
      if (typeof this.value === "number" && typeof v.value === ) {
        // small * small
      }
      // ...
    };
```
What do you think?

**Comment by koehlma** [comment]
Thank you for the long and detailed answer. I totally understand your point and exposing implementation details is indeed a bad idea. However, what do you think about using the prototype chain and introducing a new `Integer` constructor (interface)? That way `BigInteger` and `SmallInteger` remain implementation details but one can rely on integers being instances of `Integer`.

File: BigInteger.js [hunk 1]
```diff
@@ -1164,26 +1172,21 @@ var bigInt = (function (undefined) {
...
-    CACHE.randBetween = randBetween;
-    return CACHE;
+    Integer.one = Integer[1];
+    Integer.zero = Integer[0];
+    Integer.minusOne = Integer[-1];
+    Integer.max = max;
+    Integer.min = min;
+    Integer.gcd = gcd;
+    Integer.lcm = lcm;
+    Integer.isInstance = function (x) { return x instanceof BigInteger || x instanceof SmallInteger; };
```
Comment (koehlma) [inline BigInteger.js]: Can be simplified now…

**Comment by koehlma** [comment]
What do you think? I also added some tests to `spec.js`. To hide the implementation details, `bigInt(…).constructor` returns the `Integer` constructor no matter if it is a `Small`- or `BigInteger`.

**Comment by peterolson** [comment]
OK, thanks, this looks fine to me now. It even looks a bit nicer to use the `Integer` constructor than the `CACHE` object.

**Comment by koehlma** [comment]
Great, I am pleased you like it and thanks for providing and maintaining this great piece of software.

**Pull request merged**